cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java

```diff
@@ -0,0 +1,146 @@
+  }
+
+  @Nonnull
+  static List<String> getCollectionUris(@Nonnull IndexingServiceContext context,
+                                        @Nonnull CollectionFolder collectionFolder) {
+    SpelService spelService = context.getSpelService();
+    PaginatedQuery paginatedQuery = context.getPaginatedQuery();
+    RdfStoreService rdfStore = context.getRdfStore();
+    Map<String, String> queryTemplateParameters = context.getQueryTemplateParameters();
+
+    return collectionFolder.getSelectQueryResources()
+                           .stream()
+                           .map(resource -> spelService.processTemplate(resource, queryTemplateParameters))
+                           .map(query -> paginatedQuery.select(rdfStore, query))
+                           .flatMap(queryResult -> paginatedQuery.convertSingleColumnUriToStringList(queryResult)
+                                                                   .stream())
+                           .collect(Collectors.toList());
+  }
+
+  static void indexSynchronously(@Nonnull IndexingServiceContext indexingServiceContext,
+                                 @Nonnull CollectionFolder collectionFolder,
+                                 @Nonnull String indexToFill,
+                                 @Nonnull List<String> uris) {
+    List<Resource> constructQueryResources = collectionFolder.getConstructQueryResources();
+    List<Resource> facetQueryResources = collectionFolder.getFacetQueryResources();
+    for (String uri : uris) {
+      IndexMethod indexMethod = getIndexMethodForUri(indexingServiceContext, indexToFill, facetQueryResources, uri);
+      Supplier<Model> modelSupplier = getModelSupplier(indexingServiceContext, constructQueryResources, uri);
+      indexMethod.indexOne(modelSupplier, uri, IndexMethod.Configuration.SyncElasticsearch);
+    }
+  }
+
+  /**
+   * Returns <code>IndexMethod</code> instance for selected <code>index</code> and <code>collection</code>
+   *
+   * @param context             service which needs an IndexMethod instance
+   * @param indexToFill         index being filled
+   * @param facetQueryResources of collection
+   * @param uri                 of instance being indexed
+   * @return <code>IndexMethod</code> instance
+   */
+  @Nonnull
+  static IndexMethod getIndexMethodForUri(@Nonnull IndexingServiceContext context,
+                                          @Nonnull String indexToFill,
+                                          @Nonnull List<Resource> facetQueryResources,
+                                          @Nonnull String uri) {
+    Resource[] queryResources = facetQueryResources.toArray(new Resource[0]);
+    SparqlSelectToJson sparqlSelectToJson = new SparqlSelectToJson(queryResources,
+                                                                   context.getSpelService(),
+                                                                   getTemplateParameterMap(context, uri));
+    return new IndexMethod(context.getPaginatedQuery(),
```

Comment:
@natan-cox-cognizone what do you think about ading here a Function<Model, String> for externalized computable facets? for some complicated cases when we would need java to produce facet json with java?

Reply:
@alexey-lukashov Yes. For content facets, we are going to need something more pluggable. 

Also, the current setup is probably not too good. API looks a bit weird and performance is suffering because we always need to "read" Resource inputstream.

This is definitely an area which requires improvement.

Reply:
also might be useful to add timestamp for every indexed resource, and implement garbage collector based on elastic delete query, its risky but super fast deleting obsolete stuff

Reply:
@alexey-lukashov https://github.com/cognizone/asquare/issues/50